azure/src/main/java/org/apache/iceberg/azure/adlsv2/ADLSFileIO.java

```diff
@@ -0,0 +1,197 @@
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.common.DynConstructors;
+import org.apache.iceberg.io.BulkDeletionFailureException;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.io.FileInfo;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.SupportsBulkOperations;
+import org.apache.iceberg.io.SupportsPrefixOperations;
+import org.apache.iceberg.metrics.MetricsContext;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.util.SerializableMap;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** FileIO implementation backed by Azure Data Lake Storage Gen2. */
+public class ADLSFileIO implements FileIO, SupportsBulkOperations, SupportsPrefixOperations {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ADLSFileIO.class);
+  private static final String DEFAULT_METRICS_IMPL =
+      "org.apache.iceberg.hadoop.HadoopMetricsContext";
+
+  private static final HttpClient HTTP = HttpClient.createDefault();
+
+  private AzureProperties azureProperties;
+  private MetricsContext metrics = MetricsContext.nullMetrics();
+  private SerializableMap<String, String> properties = null;
+
+  /**
+   * No-arg constructor to load the FileIO dynamically.
+   *
+   * <p>All fields are initialized by calling {@link ADLSFileIO#initialize(Map)} later.
+   */
+  public ADLSFileIO() {}
+
+  @VisibleForTesting
+  ADLSFileIO(AzureProperties azureProperties) {
+    this.azureProperties = azureProperties;
+  }
+
+  @Override
+  public InputFile newInputFile(String path) {
+    return ADLSInputFile.of(path, fileClient(path), azureProperties, metrics);
+  }
+
+  @Override
+  public InputFile newInputFile(String path, long length) {
+    return ADLSInputFile.of(path, length, fileClient(path), azureProperties, metrics);
+  }
+
+  @Override
+  public OutputFile newOutputFile(String path) {
+    return ADLSOutputFile.of(path, fileClient(path), azureProperties, metrics);
+  }
+
+  @Override
+  public void deleteFile(String path) {
+    // There is no specific contract about whether delete should fail
+    // and other FileIO providers ignore failure.  Log the failure for
+    // now as it is not a required operation for Iceberg.
+    try {
+      fileClient(path).delete();
+    } catch (Exception e) {
+      LOG.warn("Failed to delete path: {}", path, e);
+    }
+  }
+
+  @Override
+  public Map<String, String> properties() {
+    return properties.immutableMap();
+  }
+
+  public DataLakeFileSystemClient client(String path) {
+    ADLSLocation location = new ADLSLocation(path);
+    return client(location);
+  }
+
+  @VisibleForTesting
+  DataLakeFileSystemClient client(ADLSLocation location) {
```

Comment:
I think we want to maintain a reference to the client after creation so it can be reused.

Reply:
The clients are basically lightweight wrappers around the HTTP client, unlike something like S3Client, and are scoped to a container or file and token/creds, so I don't feel there is much benefit in reusing or caching these.

Reply:
(The HTTP client we are reusing.)